### ISSUE TITLE
Loss my AC after update to msmart==0.1.24

### ISSUE THREAD

**Issue by tehnoinstyle** [mac-zhou/midea-ac-py#43]
**Describe the bug （描述一下问题）**
Hi, Mac.

After the last update to version msmart==0.1.24 (I updated on 12/01/2020), I get constant disappearing of my air conditioners. I see it in Logbook and History. There are no records of this in the Logs. Only once, right after the update:

```
Log Details (ERROR)
Logger: homeassistant.components.websocket_api.http.connection
Source: core.py:1405
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: December 1, 2020, 9:04:03 AM (1 occurrences)
Last logged: December 1, 2020, 9:04:03 AM

[140718739954560] string value is None for dictionary value @ data['preset_mode']
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 135, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1405, in async_call
    processed_data = handler.schema(service_data)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/validators.py", line 218, in __call__
    return self._exec((Schema(val) for val in self.validators), v)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/validators.py", line 340, in _exec
    raise e if self.msg is None else AllInvalid(self.msg, path=path)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/validators.py", line 336, in _exec
    v = func(v)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 817, in validate_callable
    return schema(data)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 594, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 432, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: string value is None for dictionary value @ data['preset_mode']
```
Before that, on version msmart==0.1.23 this was not. Everything worked fine and stable for a few months.

**Screenshots / Logs / Pcap File（屏幕截图/日志/抓包文件）**
![midea_ac_drop](https://user-images.githubusercontent.com/9967078/101241265-2b89a980-36fd-11eb-9e8c-2d02507a9d21.png)

![midea_ac_drop1](https://user-images.githubusercontent.com/9967078/101241273-33e1e480-36fd-11eb-965b-29e64f88c6e7.png)

请提交相关截图，日志，抓包让开发者能够更快的解决问题

**Versions (版本信息)**
 - Home Assistant version: 0.118.4
 - Midea msmart version: ==0.1.24
 - My air conditioners: Midea Midea Mission II MB-09N8D6H-I/MB-09N8D6H-O and Midea Mission II MB-12N8D6H-I/MB-12N8D6H-O
 - Wi-Fi module SK-102

Thanks.

**Comment by tentious** [comment]
Also experiencing same problem.

**Comment by tehnoinstyle** [comment]
Hi, Mac.

I went back to version **msmart==0.1.23** and my air conditioners are fine again. Air conditioners **do not disappear**.

**Comment by jstruggler** [comment]
Hello Tehnoinstyle,

How did you rollback to version 0.1.23?
I'm having the same behaviour, would like to downgrade but don't know how.

Regards

**Comment by tehnoinstyle** [comment]
> How did you rollback to version 0.1.23?

Hello.
In file /custom_components/midea_ac/manifest.json

```
{
        "domain": "midea_ac",
        "name": "Midea Smart Aircon",
        "documentation": "",
        "requirements": ["msmart==0.1.23", "pycryptodome", "midea"],
        "dependencies": [],
        "codeowners": ["@mac-zhou"]
}
```

**Comment by rokam** [comment]
Huummm. The real issue is that v 0.1.23 wasn't passing the right status to homeassistant, this looks like a poor engineering solution from Midea, that sometimes commands weren't available to pass to the AC or getting the status fails.

On version 0.1.23 the device remains with the last status, no matter what. I really don't know what is best.

**Comment by rokam** [comment]
I'm simply trying to ping my devices and the latency is so poor, and I also experience some drops.
![image](https://user-images.githubusercontent.com/2856518/101823981-59f6f280-3b0a-11eb-8efd-6e9b874ff64c.png)

**Comment by rokam** [comment]
Maybe the best solution is to implement it here:

`
@property
    def available(self):
        """Checks if the appliance is available for commands."""
        return self._device.online
`

Instead of getting the device status based on the commands, simply check it by pinging it. Any other ideas?

**Comment by tehnoinstyle** [comment]
> The real issue is that v 0.1.23 wasn't passing the right status to homeassistant

I did not notice such problems on version msmart==0.1.23. The integration functioned fine.
Therefore, I have returned to version msmart==0.1.23 for now.

**Comment by jstruggler** [comment]
I've returned to 0.1.23 and working fine now. I'm not noticing that devices remains with the last status.

**Comment by jstruggler** [comment]
> Maybe the best solution is to implement it here:
> 
> ```python
> @property
>     def available(self):
>         """Checks if the appliance is available for commands."""
>         return self._device.online
> ```
> 
> Instead of getting the device status based on the commands, simply check it by pinging it. Any other ideas?

Rokam,
I think the ping would be a good approach. What's the difference between version 0.1.23 and 0.1.24? something has changed on v0.1.24 that times out when checking the AC status.

I have 5 inverters and all of them became unavailable from time to time on v0.1.24.

**Comment by jstruggler** [comment]
> Maybe the best solution is to implement it here:
> 
> ```python
> @property
>     def available(self):
>         """Checks if the appliance is available for commands."""
>         return self._device.online
> ```
> 
> Instead of getting the device status based on the commands, simply check it by pinging it. Any other ideas?

Rokam,
I've just checked that this routine is present on both 0.1.23 and 0.1.24. Not sure that's what's causing the status time out.

Regards

**Comment by rokam** [comment]
Every time you receive this kind of log means that the device became offline and not available.
> 2020-12-11 11:39:49 INFO (SyncWorker_5) [msmart.lan] Couldn't connect with Device 192.168.0.47:6444
2020-12-11 11:40:04 INFO (SyncWorker_0) [msmart.lan] Couldn't connect with Device 192.168.0.47:6444
2020-12-11 11:40:19 INFO (SyncWorker_4) [msmart.lan] Couldn't connect with Device 192.168.0.47:6444
2020-12-11 11:40:34 INFO (SyncWorker_3) [msmart.lan] Couldn't connect with Device 192.168.0.47:6444

**Comment by djtef** [comment]
I agree with you @rokam , the AC is really unavailable, in 0.1.23 Home Assistant believes it is but it's false, we can't control the AC.
But I would like to understand why it turns anavailable from time to time, sometimes it can last 1 hour ! Is it a wifi dongle problem or another AC problem ?

**Comment by rokam** [comment]
I'm sorry @djtef but IDK. I think that hass.io checks from time to time if the device is available and it waits longer to check again when the device is unavailable.

**Comment by djtef** [comment]
I think the AC is really unavailable for a long time. I tried to send directly a python command line during this unavailability and I got : 
`INFO:msmart.lan:Couldn't connect with Device 192.168.0.20:6444` until it came back available in home assistant.

**Comment by rokam** [comment]
I think I had reduced the length of unavailability by creating a different wifi network at 2.4GHz only (using standard b). I'll later try at least standard g.

**Comment by djtef** [comment]
OK so you think that the problem comes from the WiFi usb key

**Comment by rokam** [comment]
@djtef I really don't know. Although I've found this evidence in Portuguese:

> Recebi ligação do Sr. Wesley Soares (MIDEA) informando que por eu ter duas redes em frequências diferentes (2.4 e 5 GHz) qualquer oscilação ocasiona o OFFLINE no aparelho de ar condicionado. Também me informou, provavelmente muito equivocadamente, que o fato do celular estar a 5GHz e o Kit WiFi em 2.4 GHz isso causa esse problema, ignorando a nuvem do Midea Air App no processo de comunicação. No site MIDEA (http://www.mideadobrasil.com.br/pt/produtos/interna/408/kit-wi-fi-da-midea) não há qualquer menção de não poder co-existir em uma residências duas redes de frequências diferentes. Em resumo, problema não solucionado pela MIDEA do Brasil.

https://www.reclameaqui.com.br/midea/kit-wi-fi-inverter-fica-offline-no-app-midea-air_yr1gARZUArRKbmex/

This says that MIDEA have informed the customer that the issue is related to coexisting 2.5GHz and 5GHz wifi networks.

**Comment by djtef** [comment]
Interisting, I have both WiFi Frequency  on my router. When I lose the network communication I have to power off and power on the AC to fix the problem. I will try to disconnect and reconnect the WiFi usb key. What I know is that restart the router is not sufficient

**Comment by rokam** [comment]
@djtef I do only disconnect and reconnect the usb dongle and it works.

**Comment by rokam** [comment]
I'm still getting some drops but usually it takes just a few seconds to reconnect:
![image](https://user-images.githubusercontent.com/2856518/102414440-4f8a9c00-3fd5-11eb-8640-ac465d58b883.png)
![image](https://user-images.githubusercontent.com/2856518/102414488-6630f300-3fd5-11eb-942f-bb34d2029a0c.png)
![image](https://user-images.githubusercontent.com/2856518/102414535-7779ff80-3fd5-11eb-8d45-886bbf30fe16.png)

**Comment by djtef** [comment]
I have exactly the same problem, but sometimes it's longer, last night the AC lost completely the network, I had to restart the AC this morning. 
The problem is that we can't disconnect the dongle remotly 😅

**Comment by haozgit** [comment]
hi...both 0.1.23 and 0.1.24 doesn't work for me. any idea? 
tried to remove and reconnect usb module as well. no luck

**Comment by haozgit** [comment]
> hi...both 0.1.23 and 0.1.24 doesn't work for me. any idea?
> tried to remove and reconnect usb module as well. no luck

just fyi, im using dongle EU-SK103

**Comment by rokam** [comment]
@haozgit that is not the same issue. I suggest you open another issue.

**Comment by jstruggler** [comment]
Rokam,
These dongles seem to restart them selfs. 
I wonder if is possible to do the following:
1- Don't show the AC's as Unavailable as of version 0.1.23
2- In case of unavailability, store the user command and resend whenever the AC is reachable again.

Regards

**Comment by rokam** [comment]
@jstruggler I don't see them only restarting themselves. I see, on a regular basis the AC losing connection with the Wi-Fi network. 

> I think I had reduced the length of unavailability by creating a different wifi network at 2.4GHz only (using standard b). I'll later try at least standard g.

That strategy didn't work, today I lost connection with 2 AC units.

**Comment by christosmilts** [comment]
Hi All,

I have 2 Midea ACs. One came with pre-installed usb dongle and works fine. On the 2nd one I've plugged a SK 103 usb and it's shown as unavailable (using msmart==0.1.24). When downgrading to msmart==0.1.23, the 2nd one seems to be available, but isn't really working - (the 1st one works still OK with 0.1.23, btw). 
Is there a compatibility rule for SK-103, SK-102 dongles?

**Comment by rokam** [comment]
@christosmilts the issue isn't from the integration. The issue is with the USB dongles. You won't be able to use your AC even with the Midea App when it shows unavailable at homeassistant

**Comment by rokam** [comment]
https://www.midea.com.ua/sites/default/files/downloadables-documents/pdf/wifi-control-en-201911.pdf

Are you kidding me Midea?
![image](https://user-images.githubusercontent.com/2856518/103390351-b0949280-4af2-11eb-97f6-6b35055d2a9d.png)
Too many devices??? What do they mean?

**Comment by christosmilts** [comment]
> @christosmilts the issue isn't from the integration. The issue is with the USB dongles. You won't be able to use your AC even with the Midea App when it shows unavailable at homeassistant

Actually it does work fine with Midea App, but it shows 'unavailable' in HA: 
![image](https://user-images.githubusercontent.com/67158261/103404540-9a232100-4b5c-11eb-8759-f33d97c0b495.png)

**Comment by rokam** [comment]
@christosmilts can you even change the AC status in the Midea App? When this happens to me, I'm unable to change any settings from the AC. FYI it shows unavailable when the appliance loose connectivity with the WIFI network

**Comment by jstruggler** [comment]
Hello,
I think the best approach is to ignore the temporary unavailability and to store the user commands to resend whenever the device is back online. The command queue should have a time out of a few seconds for long term or intencional offline periods.

That’s the reason Im using version smart 0.1.23 that doesn’t make it unavailable on Home Assistant when the devices goes offline. I have five AC’s, my AC Lovelace page is always changing states on version 0.1.24, it’s very annoying.
I’ve increased devices stability with wifi reinforcement on the installation spots, but got offline events even with near ideal wifi conditions.

Regards

**Comment by ihaettypo** [comment]
Exactly the same problem here as well!

**Comment by Asom-Velz** [comment]
> > hi...both 0.1.23 and 0.1.24 doesn't work for me. any idea?
> > tried to remove and reconnect usb module as well. no luck
> 
> just fyi, im using dongle EU-SK103

Dongle OSK-103 is more stable with midea app and you don’t have the unavailability problems but is not recognized from mac-zhou/midea-ac-py. I just open an issue for that.

**Comment by mac-zhou** [comment]
I specifically created a 2.4G wireless network for iot devices to use, iot devices generally only support 2.4G wireless networks, often with problems, not just midea.
So it is recommended to separate 2.4G and 5G.

**Comment by Asom-Velz** [comment]
I don't have 5G - only 2.4G and I faced the same unavailability problems with OSK102 module

**Comment by mac-zhou** [comment]
> I don't have 5G - only 2.4G and I faced the same unavailability problems with OSK102 module

Have you tried increasing the scan interval? default is 15s.
You can add `scan_interval: 60` to the configure file and see if it still happens.

**Comment by mac-zhou** [comment]
or i add a option configure to force online?

**Comment by mac-zhou** [comment]
@rokam I think I'm ready to roll back to v0.1.23, there are too many problems with UNAVAILABLE and this is causing problems for many people. What do you think?

**Comment by Asom-Velz** [comment]
It would be good to roll back to v0.1.23 but it would be better to implement the new OSK103 dongle since the OSK102 is obsolete and we can’t find it anymore in the market

**Comment by rokam** [comment]
@mac-zhou its up to you. I think that is better to know when the device is unavailable. Although I think that a config option it's a better approach, cause each one chooses what is best for you.

**Comment by rokam** [comment]
I've made a PR #69 to add that option but it depends on the msmart PR as well.

**Comment by mac-zhou** [comment]
> It would be good to roll back to v0.1.23 but it would be better to implement the new OSK103 dongle since the OSK102 is obsolete and we can’t find it anymore in the market

The V3 protocol is no longer purely a LAN protocol, the encryption key is required to be taken interactively with the cloud.
Other developers use [facke cloud](https://github.com/WMP/midea-ac-py/tree/support-8370#install-fake-cloud-only-if-you-have-protocol-version-3) or [midea-open-dongle](https://github.com/dudanov/midea-open-dongle) to solve this problem.
but, the point of this project is not to abandon the midea cloud or replace hardware.
